pkg/spec3/benchmark_serialization_test.go

```diff
@@ -105,16 +106,16 @@ func BenchmarkOpenAPIV3Deserialize(b *testing.B) {
 			}
 		})
 
-		// TODO: Enable this benchmark when jsonv2 is functional for OpenAPI V3.
-		// b.Run("jsonv2", func(b2 *testing.B) {
-		// b2.ReportAllocs()
-		// 	internal.UseOptimizedJSONUnmarshaling = true
-		// 	for i := 0; i < b2.N; i++ {
-		// 		var result OpenAPI
-		// 		if err := result.UnmarshalJSON(originalJSON); err != nil {
-		// 			b2.Fatal(err)
-		// 		}
-		// 	}
-		// })
+		b.Run("jsonv2", func(b2 *testing.B) {
+			b2.ReportAllocs()
+			internal.UseOptimizedJSONUnmarshaling = true
+			internal.UseOptimizedJSONUnmarshalingV3 = true
+			for i := 0; i < b2.N; i++ {
+				var result *OpenAPI
+				if err := jsonv2.Unmarshal(originalJSON, &result); err != nil {
```

Comment:
Thanks for bringing this up, yeah external callers shouldn't be able to access the internal json Unmarshal and a custom Unmarshal function should be implemented. I was having some trouble implementing the `UnmarshalJSON` function because we want to switch to jsonv2 if a flag is enabled, but continue with jsonv1 otherwise. Implementing a custom unmarshaler bypasses the default jsonv1 unmarshalling and I didn't find an easy way to call it (Ideally something like a `MarshalNoCustom()` kind of function similar to the `FuzzNoCustom()` in gofuzz). The temporary workaround is to name the function something other than `UnmarshalJSON` (currently set as `CustomUnmarshalJSON`) though I'm not really a fan of doing it this way. I'll keep looking into this

```
func (o *OpenAPI) CustomUnmarshalJSON(data []byte) error {
    if internal.UseOptimizedJSONUnmarshalingV3 {
        return jsonv2.Unmarshal(data, &o)
    }
    return json.Unmarshal(data, &o)
}
```

Reply:
https://github.com/kubernetes/kube-openapi/blob/39b3636cbaa32f381a5e7090cda15c71f38dd993/pkg/validation/spec/tag.go#L56-L65

If the flag is on, then we switch to jsonv2.

Otherwise, we call the jsonv1 version on all the embedded structs. Embedded structs were things like `TagProps` which do not have custom marshaling functions.

Reply:
The prerequisite is the struct must be an embedded struct though right? For a non embedded struct, this unmarshal process would have to be repeated on all fields? https://github.com/kubernetes/kube-openapi/blob/master/pkg/spec3/spec.go#L24-L37

Reply:
Here is a trick that can work for the case you linked:

```go
func (o *OpenAPI) UnmarshalJSON(data []byte) error {
        if internal.UseOptimizedJSONUnmarshaling { 
 		return jsonv2.Unmarshal(data, t) 
 	} 
	type OpenAPIWithNoFunctions OpenAPI
	var me *OpenAPIWithNoFunctions = (*OpenAPIWithNoFunctions)(o)
	return json.Unmarshal(data, me)
}
```

I think this might have a penalty performing syntax validation on the input twice.

Passed this test:

```go
func TestUnmarshalOpenAPI(t *testing.T) {
	actual := &spec3.OpenAPI{}
	text := `{"openapi":"3.0.0", "info": {"title":"Hello, world!"}}`
	expect := &spec3.OpenAPI{
		Version: "3.0.0",
		Info: &spec.Info{
			InfoProps: spec.InfoProps{
				Title: "Hello, world!",
			},
		},
	}

	require.NoError(t, json.Unmarshal([]byte(text), actual))
	require.Equal(t, expect, actual)
}
```